examples/neighbors/approximate_nearest_neighbors.py

```diff
@@ -121,191 +100,209 @@ def transform(self, X):
+from sklearn.pipeline import make_pipeline
 
-    nms = NMSlibTransformer()
-    Xt2 = nms.fit_transform(X)
+datasets = [
+    ("MNIST_10000", load_mnist(n_samples=2000)),
+    ("MNIST_20000", load_mnist(n_samples=20000)),
+]
+
+n_iter = 500
+perplexity = 30
+metric = "euclidean"
+# TSNE requires a certain number of neighbors which depends on the
+# perplexity parameter.
+# Add one since we include each sample as its own neighbor.
+n_neighbors = int(3.0 * perplexity + 1) + 1
+
+tsne_params = dict(
+    init="random",  # pca not supported for sparse matrices
+    perplexity=perplexity,
+    method="barnes_hut",
+    random_state=42,
+    n_iter=n_iter,
+    learning_rate="auto",
+)
+
+transformers = [
+    (
+        "KNeighborsTransformer",
+        KNeighborsTransformer(n_neighbors=n_neighbors, mode="distance", metric=metric),
+    ),
+    (
+        "NMSlibTransformer",
+        NMSlibTransformer(n_neighbors=n_neighbors, metric=metric),
+    ),
+    (
+        "PyNNDescentTransformer",
+        PyNNDescentTransformer(
+            n_neighbors=n_neighbors, metric=metric, parallel_batch_queries=True
+        ),
+    ),
+]
 
-    assert_array_almost_equal(Xt0.toarray(), Xt1.toarray(), decimal=5)
-    assert_array_almost_equal(Xt0.toarray(), Xt2.toarray(), decimal=5)
+for dataset_name, (X, y) in datasets:
 
+    msg = "Benchmarking on %s:" % dataset_name
+    print("\n%s\n%s" % (msg, "-" * len(msg)))
 
-def load_mnist(n_samples):
-    """Load MNIST, shuffle the data, and return only n_samples."""
-    mnist = fetch_openml("mnist_784", as_frame=False, parser="pandas")
-    X, y = shuffle(mnist.data, mnist.target, random_state=2)
-    return X[:n_samples] / 255, y[:n_samples]
+    for transformer_name, transformer in transformers:
+        longest = np.max([len(name) for name, model in transformers])
+        whitespaces = " " * (longest - len(transformer_name))
+        for _ in range(2):
+            start = time.time()
+            transformer.fit(X)
+            fit_duration = time.time() - start
+            print(
+                "%s: %s%.3f sec (fit)" % (transformer_name, whitespaces, fit_duration)
+            )
+        for _ in range(2):
+            start = time.time()
+            Xt = transformer.transform(X)
+            transform_duration = time.time() - start
+            print(
+                "%s: %s%.3f sec (transform)"
+                % (transformer_name, whitespaces, transform_duration)
+            )
+
+# %%
+# Sample output::
+#
+#     Benchmarking on MNIST_10000:
+#     ----------------------------
+#     KNeighborsTransformer:  0.005 sec (fit)
+#     KNeighborsTransformer:  0.004 sec (fit)
+#     KNeighborsTransformer:  1.285 sec (transform)
+#     KNeighborsTransformer:  1.162 sec (transform)
+#     NMSlibTransformer:      0.226 sec (fit)
+#     NMSlibTransformer:      0.235 sec (fit)
+#     NMSlibTransformer:      0.323 sec (transform)
+#     NMSlibTransformer:      0.295 sec (transform)
```

Comment:
I think there is no point in running those several times because they don't have the cold start problem of `PyNNDescentTransformer` which comes from the numba jit.